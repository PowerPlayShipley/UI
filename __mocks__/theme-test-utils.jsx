// TODO: Fix this??

import * as React from 'react'
import userEvent from '@testing-library/user-event'
import { render as rtlRender } from '@testing-library/react'

import { ThemeProvider } from "styled-components";

import theme from '../src/themes'
import GlobalStyle from "../src/components/GlobalStyle";

function renderWithTheme(ui, { ...options } = {}) {
  const Wrapper = ({ children }) => (
    <ThemeProvider theme={theme}>
      <GlobalStyle />
      {children}
      <div id='modal-root' />
    </ThemeProvider>
  )

  return rtlRender(ui, { wrapper: Wrapper, ...options })
}

// THE BELLOW COME FROM `https://github.com/styled-components/styled-components/blob/0aa3170c255a49cd41c3fbeb2b8051b5d132f229/src/test/utils.js`

const stripComments = (str) => str.replace(/\/\*.*?\*\/\n?/g, '');

const stripWhitespace = (str) => str
  .trim()
  .replace(/([;\{\}])/g, '$1  ')
  .replace(/\s+/g, ' ');

const getCSS = (scope) =>
  Array.from(scope.querySelectorAll('style'))
    .map(tag => tag.innerHTML)
    .join('\n')
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

function expectCSSMatches ( _expectation, opts = { ignoreWhitespace: true }) {
  // NOTE: This should normalise both CSS strings to make irrelevant mismatches less likely
  const expectation = _expectation
    .replace(/ {/g, '{')
    .replace(/:\s+/g, ':')
    .replace(/:\s+;/g, ':;');

  const css = getCSS(document);

  if (opts.ignoreWhitespace) {
    const stripped = stripWhitespace(stripComments(css));
    expect(stripped).toEqual(stripWhitespace(expectation));
    return stripped;
  } else {
    expect(css).toEqual(expectation);
    return css;
  }
};

export * from '@testing-library/react';

export { renderWithTheme, expectCSSMatches };
