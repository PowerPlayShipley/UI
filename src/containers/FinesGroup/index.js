/**
 * FinesGroup is the handler for each fine's section i.e. the players/captains.
 *
 * It has to be placed inside a `ConfigProvider` this will be an auto updating configuration handler.
 * The remaining code should be able to be used in isolation, HOC to handle the updates maybe ?? Where
 * it can update its own fines data and no one else.
 * */

import React, { useRef, useCallback, useState } from 'react'
import PropTypes from 'prop-types'

import Wrapper from "./Wrapper";

import FinesTable from "../FinesTable";
import Header from "../../components/Header";
import Button from "../../components/Button";
import { ModalSelector } from "../../components/Selector";

// Just set this, so we can update the right value in the database
const getCellFromArgs = (index, value, { index: idx }, { id }) => {
  return { row: idx, col: parseInt(id.slice(-1)) }
}

const FinesGroup = ({ data, configuration, title, onFineAdded, onFineRemoved, ...rest }) => {
  const selector = useRef(null)

  // Add state management
  const [cell, setCell] = useState(null) // { row: x, col: x }

  // Handle any work when the selector has a new item
  const handleSelectorClicked = useCallback((item, _) => {
    selector.current.close()

    // Pass this up the chain
    onFineAdded && onFineAdded(item, cell)

    setCell(null)
  }, [selector, onFineAdded, cell, setCell])

  // Handle any work when an old fine has been clicked
  const handleItemClicked = useCallback((index, value, row, col) => {
    const cell = getCellFromArgs(index, value, row, col)
    onFineRemoved && onFineRemoved(index, value[index], cell)
  }, [onFineRemoved])

  // Add any rendering for the selector here. Keep it cleaner
  const renderSelector = (style) => (
    <ModalSelector style={style} title='Select Fines' defaultOpen={false} onItemClicked={handleSelectorClicked}
                   onCloseClicked={() => { selector.current.close(); setCell(null) }} ref={selector} items={configuration} />
  )

  return (
    <Wrapper {...rest}>
      <div className='meta'>
        <Header level='h2' title={title} />
        <div className='btn-group'>
          <Button title='Edit' state='primary' />
        </div>
      </div>
      <FinesTable data={data} onToolbarClick={(...args) => { setCell(getCellFromArgs(...args)); selector.current.open() }} onItemClick={handleItemClicked} />
      {renderSelector({ alignSelf: 'center', margin: '8px auto', minWidth: '200px' })}
    </Wrapper>
  )
}

FinesGroup.propTypes = {
  // Will set this properly when I know exactly where it comes from and who manipulates this from the API response
  data: PropTypes.array.isRequired,
  configuration: PropTypes.array,
  title: PropTypes.string.isRequired,
  onFineAdded: PropTypes.func,
  onFineRemoved: PropTypes.func
}

export default FinesGroup
