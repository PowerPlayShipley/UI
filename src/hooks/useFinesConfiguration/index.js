import { useMemo } from "react";

const useFinesConfiguration = (initial) => {
  const finesConfiguration = useMemo(() => Object.keys(initial.fines).map((key) => ({
    value: key,
    tooltip: initial.fines[key].name
  })), [initial])

  return ({ finesConfiguration, config: initial })
}

export default useFinesConfiguration
