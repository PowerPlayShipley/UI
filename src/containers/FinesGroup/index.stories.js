import React from 'react'
import { action } from "@storybook/addon-actions";

import FinesGroup from "./index";
import useFinesConfiguration from "../../hooks/useFinesConfiguration";

export default {
  title: 'Fines/Group',
  component: FinesGroup
}

const Template = ({ configuration, data, ...args }) => {
  const { finesConfiguration } = useFinesConfiguration(configuration)
  return (
    <div style={{ backgroundColor: '#f5f5f5', padding: '16px 16px' }}>
      <FinesGroup configuration={finesConfiguration} data={data} {...args} />
    </div>
  )
};

export const Default = Template.bind({});
Default.args = {
  title: 'Players',
  configuration: {
    players: {
      'HW': {
        name: 'Harry'
      },
      'SK': {
        name: 'Ste'
      }
    },
    fines: {
      H: {
        name: 'Hanging',
        cost: 50
      },
      D: {
        name: 'Diagonal Hanging',
        cost: 50
      },
      R: {
        name: 'Reverse',
        cost: 50
      }
    }
  },
  data: [
    {
      player: 'Ade',
      games: [["H"], ["D", "H", "D", "D"], ["R"]]
    },
    {
      player: 'Tom Collins',
      games: [[], ["H"], ["R"]]
    },
    {
      player: 'Jamie',
      games: [["H", "D", "H", "D", "D"], ["D", "H"], ["R"]]
    }
  ],
  onFineAdded: action('onFineAdded'),
  onFineRemoved: action('onFineRemoved'),
}

