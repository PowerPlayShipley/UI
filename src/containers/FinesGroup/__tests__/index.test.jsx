/**
 * @jest-environment jsdom
 */

import React from 'react'
import { renderHook } from '@testing-library/react-hooks'
import { renderWithTheme, fireEvent } from '../../../../__mocks__/theme-test-utils'

import FinesGroup from "../index";
import useFinesConfiguration from "../../../hooks/useFinesConfiguration";

const configuration = {
  players: {
    'AA': {
      name: 'Ade'
    }
  },
  fines: {
    H: {
      name: 'Hanging',
      cost: 50
    },
    R: {
      name: 'Reverse',
      cost: 50
    }
  }
}

const data = [
  {
    player: 'AA',
    games: [["H"]]
  }
]

describe('Display FinesGroup', function () {
  it('should create a valid snapshot', function () {
    const { result } = renderHook(() => useFinesConfiguration(configuration))

    const { container } = renderWithTheme(<FinesGroup title='Players' data={data} configuration={result.current.finesConfiguration} />)
    expect(container.firstChild).toMatchSnapshot()
  });

  it('should create valid buttons', function () {
    const handleClick = jest.fn()

    const { result } = renderHook(() => useFinesConfiguration(configuration))

    const { getByText } = renderWithTheme(<FinesGroup title='Players' data={data} onFineRemoved={handleClick} configuration={result.current.finesConfiguration} />)

    fireEvent.click(getByText('H').parentNode)
    expect(handleClick).toHaveBeenCalledTimes(1)
  });

  it('should create valid toolbar', function () {
    const handleClick = jest.fn()

    const { result } = renderHook(() => useFinesConfiguration(configuration))

    const { container, getByText } = renderWithTheme(<FinesGroup title='Players' data={data} onFineAdded={handleClick} configuration={result.current.finesConfiguration} />)

    fireEvent.click(container.querySelector('svg'))
    fireEvent.click(getByText('R').parentNode)

    expect(handleClick).toHaveBeenCalledTimes(1)
  });
})
