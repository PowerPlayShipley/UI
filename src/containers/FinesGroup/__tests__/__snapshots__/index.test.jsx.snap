// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Display FinesGroup should create a valid snapshot 1`] = `
.c2 {
  position: relative;
  display: inline-block;
  padding: 5px 16px;
  font-size: 14px;
  font-weight: 600;
  line-height: 20px;
  white-space: nowrap;
  vertical-align: middle;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  border: 1px solid;
  border-radius: 6px;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
}

.c2:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c2:disabled,
.c2.disabled,
.c2[aria-disabled=true] {
  cursor: default;
}

.c2 .dropdown-caret {
  margin-left: 4px;
  opacity: 0.8;
}

.c2 .btn-icon {
  margin-right: 4px;
  color: inherit;
  vertical-align: text-bottom;
}

.c2 .btn-icon:only-child {
  margin-right: 0;
}

.c2 {
  color: #24292f;
  background-color: #f6f8fa;
  border: 1px rgba(208,215,222,0.5) solid;
  box-shadow: 0 1px 0 rgba(27,31,36,0.04),inset 0 1px 0 rgba(255,255,255,0.25);
  -webkit-transition: 0.2s cubic-bezier(0.3,0,0.5,1);
  transition: 0.2s cubic-bezier(0.3,0,0.5,1);
  -webkit-transition-property: color,background-color,border-color;
  transition-property: color,background-color,border-color;
}

.c2:hover,
.c2.hover,
[open] > .c2 {
  background-color: #eaeef2;
  -webkit-transition-duration: 0.1s;
  transition-duration: 0.1s;
}

.c2:active {
  background-color: #e1e6ec;
  -webkit-transition: none;
  transition: none;
}

.c2.selected,
.c2[aria-selected=true] {
  background-color: #e4e9ee;
  box-shadow: inset 0 1px 0 rgba(208,215,222,0.2);
}

.c2:disabled,
.c2.disabled,
.c2[aria-disabled=true] {
  color: #8c959f;
  background-color: #f6f8fa;
  cursor: default;
}

.c2:disabled .btn-icon,
.c2.disabled .btn-icon,
.c2[aria-disabled=true] .btn-icon {
  color: #8c959f;
}

.c2:focus,
.c2.focus {
  outline: none;
  box-shadow: 0 0 0 3px rgba(9,105,218,0.15);
}

.c2.primary {
  color: #ffffff;
  background-color: #1a7f37;
  border-color: rgba(208,215,222,0.5);
  box-shadow: 0 1px 0 rgba(27,31,36,0.1), inset 0 1px 0 rgba(255,255,255,0.3);
}

.c2.primary:hover,
.c2.primary.hover,
[open] > .c2.primary {
  background-color: #116329;
  border-color: rgba(208,215,222,0.5);
}

.c2.primary:active,
.c2.primary.selected,
.c2.primary[aria-selected=true] {
  background-color: #105a25;
  box-shadow: inset 0 1px 0 rgba(0,45,17,0.2);
}

.c2.primary:disabled,
.c2.primary.disabled,
.c2.primary[aria-disabled=true] {
  color: rgba(255,255,255,0.8);
  background-color: #94d3a2;
  border-color: rgba(208,215,222,0.5);
}

.c2.primary:disabled .btn-icon,
.c2.primary.disabled .btn-icon,
.c2.primary[aria-disabled=true] .btn-icon {
  color: #8c959f;
}

.c2.primary:focus,
.c2.primary.focus {
  background-color: #2da44e;
  border-color: rgba(208,215,222,0.5);
  box-shadow: 0 0 0 3px rgba(45,164,78,0.4);
}

.c2.primary .btn-icon {
  color: #ffffff;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c0 .meta {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  margin-bottom: 8px;
}

.c0 .btn-group {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 .btn-group .c1 {
  margin-right: 8px;
}

.c0 .btn-group .c1:last-child {
  margin-right: 0;
}

.c3 {
  display: block;
  width: 100%;
  overflow-x: auto;
}

.c3 table {
  width: 100%;
  max-width: 100%;
}

.c3 table .editable div {
  -webkit-flex-wrap: nowrap;
  -ms-flex-wrap: nowrap;
  flex-wrap: nowrap;
}

.c3 table .editable svg {
  color: #6e7781;
}

.c3 table {
  border: none;
  height: auto;
  border-collapse: separate;
  border-spacing: 0 8px;
}

.c3 table th,
.c3 table td {
  width: 1%;
  vertical-align: middle !important;
}

.c3 table th {
  border: none;
  border-right: 1px solid rgba(208,215,222,0.5) !important;
}

.c3 table th:last-child,
.c3 table td:last-child {
  border-right: none !important;
}

.c3 table td {
  border: none;
}

.c3 table thead {
  font-weight: 500;
}

.c3 table thead th:first-child {
  text-align: left;
}

.c3 table tbody tr {
  background-color: #ffffff;
  box-shadow: 0 3px 6px rgba(140,149,159,0.15),inset 0 1px 0 rgba(255,255,255,0.25);
}

.c3 table tbody {
  --border-radius-cell: 6px;
}

.c3 table tbody tr td:first-child {
  border-top-left-radius: var(--border-radius-cell);
  border-bottom-left-radius: var(--border-radius-cell);
}

.c3 table tbody tr td:last-child {
  border-top-right-radius: var(--border-radius-cell);
  border-bottom-right-radius: var(--border-radius-cell);
}

.c4 {
  border-spacing: 0;
  border: 1px solid #d0d7de;
}

.c4 tr:last-child td {
  border-bottom: 0;
}

.c4 th,
.c4 td {
  margin: 0;
  padding: 0.5rem;
  border-bottom: 1px solid #d0d7de;
  border-right: 1px solid #d0d7de;
}

.c4 th:last-child,
.c4 td:last-child {
  border-right: 0;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  padding: 0 0;
}

.c5 .editable {
  overflow: auto;
}

.c5 .toolbar {
  -webkit-flex: none;
  -ms-flex: none;
  flex: none;
  box-shadow: 0 1px 0 rgba(27,31,36,0.04);
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c6 div[disabled],
.c6 div.disabled,
.c6 div[aria-disabled=true] {
  pointer-events: none;
  color: #8c959f;
  opacity: 65%;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  height: 24px;
  min-width: 24px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border-radius: 3px;
  padding: 4px;
  color: #222f3e;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c8:hover {
  background-color: #dee0e2;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  max-height: 200px;
  overflow-x: hidden;
  overflow-y: auto;
  padding: 0;
}

@media (min-width:767px) {
  .c3 table tbody td .toolbar {
    visibility: hidden;
    opacity: 0;
    -webkit-transition: visibility 0s,opacity 0.1s linear;
    transition: visibility 0s,opacity 0.1s linear;
  }

  .c3 table tbody td:hover .toolbar {
    visibility: visible;
    opacity: 1;
  }
}

<div
  class="c0"
>
  <div
    class="meta"
  >
    <h2
      class="h2"
    >
      Players
    </h2>
    <div
      class="btn-group"
    >
      <button
        class="c1 c2 primary"
      >
        Edit
      </button>
    </div>
  </div>
  <div
    class="c3"
  >
    <table
      class="c4 table"
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colspan="1"
            role="columnheader"
          >
            Player
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            Game 1
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          role="row"
        >
          <td
            role="cell"
          >
            AA
          </td>
          <td
            role="cell"
          >
            <div
              class="c5"
              data="[object Object]"
              headers="[object Object],[object Object]"
            >
              <div
                class="c6 editable"
              >
                <div
                  class="c7"
                >
                  <div
                    aria-disabled="false"
                    class="c8"
                    data-item="H"
                  >
                    <div
                      class="c9"
                    >
                      H
                    </div>
                  </div>
                  <div
                    aria-disabled="false"
                    class="c8 toolbar"
                    data-class="toolbar"
                    data-item="toolbar"
                  >
                    <div
                      class="c9"
                    >
                      <svg
                        aria-hidden="true"
                        class="svg-inline--fa fa-plus fa-w-14 "
                        data-icon="plus"
                        data-prefix="fas"
                        focusable="false"
                        role="img"
                        viewBox="0 0 448 512"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                          fill="currentColor"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
